### PR TITLE
fix: Custom Migration to clean no input values as null in Daily Rounds table

### PR DESCRIPTION
**Pull request by aeswibon**
# Bug Fix

## Proposed Changes

- [x] Migrated all fields of `Daily Rounds` having zero value to `null` 
- [x] Migrated all blood pressure fields that are less than or equal to 0 to null

## Associated Issue

- Fixes coronasafe/care#1810
- Part of coronasafe/care_fe#6985

## Merge Checklist

- [x] Tests added/fixed
- [ ] Update docs in `/docs`
- [x] Linting Complete
- [ ] Any other necessary step

_*Only PR's with test cases included and passing lint and test pipelines will be reviewed*_

@coronasafe/care-backend-maintainers @coronasafe/care-backend-admins

### CODE REVIEW

**Comment by aeswibon** [comment]
> @aparnacoronasafe, Can you confirm whether zero can be a valid value in blood pressure fields?

https://github.com/coronasafe/care/issues/1810#issuecomment-1902106254

The tasks are done just waiting for the confirmation from Aparna. Once the reply comes, then I will make the PR ready.

**Comment by aeswibon** [comment]
@sainak Can you check the PR?

File: care/facility/migrations/0408_alter_dailyround_consciousness_level_and_more.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+        if obj.insulin_intake_frequency == 0:
+            obj.insulin_intake_frequency = None
+        if obj.bp and obj.bp.get("systolic") and obj.bp.get("diastolic"):
+            if obj.bp.get("systolic") < 0 and obj.bp.get("diastolic") < 0:
+                obj.bp = {}
+            else:
+                if obj.bp.get("systolic") < 0:
+                    obj.bp.update({"systolic": None})
+                if obj.bp.get("diastolic") < 0:
+                    obj.bp.update({"diastolic": None})
```
Comment (sainak) [inline care/facility/migrations/0408_alter_dailyround_consciousness_level_and_more.py:42]: use individual queries for each field
https://docs.djangoproject.com/en/4.2/topics/db/queries/#querying-jsonfield

**Review by sainak — CHANGES_REQUESTED** [review]

File: care/facility/migrations/0408_alter_dailyround_consciousness_level_and_more.py (hunk 1 above)
Comment (aeswibon) [reply care/facility/migrations/0408_alter_dailyround_consciousness_level_and_more.py]: Done

File: care/facility/migrations/0414_alter_dailyround_consciousness_level_and_more.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    DailyRound.objects.filter(insulin_intake_frequency=0).update(
+        insulin_intake_frequency=None
+    )
+    DailyRound.objects.filter(bp__systolic__lt=0, bp__diastolic__lt=0).update(bp={})
+    DailyRound.objects.filter(bp__contains={"systolic": {"$lt": 0}}).update(
+        bp=RawSQL("jsonb_set(bp, '{systolic}', 'null')", [])
+    )
+    DailyRound.objects.filter(bp__contains={"diastolic": {"$lt": 0}}).update(
+        bp=RawSQL("jsonb_set(bp, '{diastolic}', 'null')", [])
+    )
```
Comment (sainak) [inline care/facility/migrations/0414_alter_dailyround_consciousness_level_and_more.py:37]:
```suggestion
    DailyRound.objects.filter(Q(bp__systolic__lt=0)| Q(bp__diastolic__lt=0)).update(bp={"systolic":None, "diastolic": None})
```
does this work?

Comment (aeswibon) [reply care/facility/migrations/0414_alter_dailyround_consciousness_level_and_more.py]: Done

**Comment by aeswibon** [comment]
@sainak Can you check the PR?

File: care/facility/models/daily_round.py [hunk 3]
```diff
@@ -504,18 +505,29 @@ def set_push_score(item):
...
-        self.total_intake_calculated += sum([x["quantity"] for x in self.iv_fluids])
-        self.total_intake_calculated += sum([x["quantity"] for x in self.feeds])
-
-        self.total_output_calculated = sum([x["quantity"] for x in self.output])
-
-        # self.pressure_sore = self.update_pressure_sore()
+        if (
+            self.glasgow_eye_open is None
+            and self.glasgow_motor_response is None
+            and self.glasgow_verbal_response is None
```
Comment (sainak) [inline care/facility/models/daily_round.py:511]: Are these fields always compounded?
is there any case for only some of the fields used to calculate the score?

File: care/facility/models/daily_round.py [hunk 4]
```diff
@@ -504,18 +505,29 @@ def set_push_score(item):
...
+        ):
+            self.glasgow_total_calculated = None
+        else:
+            self.glasgow_total_calculated = (
+                self.cztn(self.glasgow_eye_open)
+                + self.cztn(self.glasgow_motor_response)
+                + self.cztn(self.glasgow_verbal_response)
+            )
```
Comment (sainak) [inline care/facility/models/daily_round.py:519]:
```suggestion
        if (
            self.glasgow_eye_open is not None
            and self.glasgow_motor_response is not None
            and self.glasgow_verbal_response is not None
        ):
            self.glasgow_total_calculated = (
                self.glasgow_eye_open
                + self.glasgow_motor_response
                + self.glasgow_verbal_response
            )
```
as discussed in the call glasgow total score should only be calculated if all three values are present

File: care/facility/models/daily_round.py [hunk 5]
```diff
@@ -504,18 +505,29 @@ def set_push_score(item):
...
+            self.total_intake_calculated = None
+        else:
+            self.total_intake_calculated = sum([x["quantity"] for x in self.infusions])
+            self.total_intake_calculated += sum([x["quantity"] for x in self.iv_fluids])
+            self.total_intake_calculated += sum([x["quantity"] for x in self.feeds])
+
+        if not self.output:
+            self.total_output_calculated = None
+        else:
+            self.total_output_calculated = sum([x["quantity"] for x in self.output])
```
[the remark this answers is not in the document]
Comment (aeswibon) [reply care/facility/models/daily_round.py:530]: No, they are storing 0 as a value. That's why I added the check.

[the remark this answers is not in the document]
Comment (aeswibon) [reply care/facility/models/daily_round.py:530]: Done

File: care/facility/models/daily_round.py (hunk 4 above)
Comment (aeswibon) [reply care/facility/models/daily_round.py:519]: Done

**Review by sainak — APPROVED** [review]
Lgtm

**Review by Ashesh3 — APPROVED** [review]

File: care/facility/migrations/0415_alter_dailyround_consciousness_level_and_more.py [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+    DailyRound.objects.filter(rhythm=0).update(rhythm=None)
+    DailyRound.objects.filter(ventilator_mode=0).update(ventilator_mode=None)
+    DailyRound.objects.filter(ventilator_interface=0).update(ventilator_interface=None)
+    DailyRound.objects.filter(ventilator_oxygen_modality=0).update(
+        ventilator_oxygen_modality=None
+    )
+    DailyRound.objects.filter(insulin_intake_frequency=0).update(
+        insulin_intake_frequency=None
+    )
+    DailyRound.objects.filter(Q(bp__systolic__lt=0) | Q(bp__diastolic__lt=0)).update(
```
Comment (vigneshhari) [inline care/facility/migrations/0415_alter_dailyround_consciousness_level_and_more.py:31]: Possible loss of data?

Comment (aeswibon) [reply care/facility/migrations/0415_alter_dailyround_consciousness_level_and_more.py:31]: We need to correct the value as both of them should be greater than -1. If either of them has invalid data, then we need to convert it into an empty object. This has already been discussed in previous EODs.
cc: @sainak

**Review by vigneshhari — APPROVED** [review]

**Pull request merged**